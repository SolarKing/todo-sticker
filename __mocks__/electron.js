// @flow
let listeners: Map<string, Function> = new Map();

class IpcMain {   

    on(channel: string, listener: Function) {
        listeners.set(channel, listener);
    }
    
}

class IpcRenderer {   

    send(channel: string, ...args: Array<mixed>) {
        const listener = listeners.get(channel);
        if (listener != null) {
            listener({}, ...args);
        }
    }
    
}

class BrowserWindow {
    _listeners: Map<string, Function> = new Map();

    webContents = {
        openDevTools() {

        },
        on(channel: string, listener: Function) {
            listeners.set(channel, listener);
        },
        emit(channel: string, ...args: Array<mixed>) {
            const listener = listeners.get(channel);
            if (listener != null) {
                listener(...args);
            }
        },
        destroy() {

        }
    }

    constructor() {
        this._listeners = new Map();
    }
    isMaximized() {
        return false;
    }
    loadURL() { }
    close() { }
    maximize() { }
    unmaximize() { }
    minimize() { }

    on(channel: string, listener: Function) {
        this._listeners.set(channel, listener);
    }

    emit(channel: string, ...args: Array<mixed>) {
        const listener = this._listeners.get(channel);
        if (listener != null) {
            listener(...args);
        }
    }

    static getFocusedWindow(): BrowserWindow {
        return new BrowserWindow();
    }

}

const ipcMain = new IpcMain();
const ipcRenderer = new IpcRenderer();

export {
    ipcMain,
    ipcRenderer,
    BrowserWindow
};