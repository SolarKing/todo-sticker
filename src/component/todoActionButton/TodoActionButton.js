// @flow
import React from 'react';
import { Icon } from 'antd';
import { Menu, Dropdown, Button } from 'antd';
import EditModal from '../../container/editTodoModalContainer';
import Translate from '../../class/translate';
import type {TodoState} from '../../states/index';
import DeleteButtonContainer from '../../container/DeleteButtonContainer';

type Props = {
    todo: TodoState,
}

type States = {
    editModalVisible: boolean,
}

export default class TodoActionButton extends React.Component<Props, States> {

    handleEdit: Function;

    state: States = {
        editModalVisible: false,
    }

    constructor(props: Props) {
        super(props);
        this.handleEdit = this.handleEdit.bind(this);
    }

    handleEdit() {
        this.setState({
            editModalVisible: true
        });
    }

    render() {
        let {
            todo,
            ...others
        } = this.props;

        const menu = (
            <Menu>
                <Menu.Item>
                    <a>Complete</a>
                </Menu.Item>
                <Menu.Item>
                    <a onClick={this.handleEdit} >Edit</a>
                </Menu.Item>
                <Menu.Item>
                    <DeleteButtonContainer todoId={todo.id} />
                </Menu.Item>
            </Menu>
        );
        return (
            <div {...others}>
                <Dropdown overlay={menu} trigger={['click']}>
                    <Button><Icon type="ellipsis" /></Button>
                </Dropdown>
                <EditModal
                    title={Translate.tr('Edit Todo')}
                    visible={this.state.editModalVisible}
                    onSave={() => this.setState({editModalVisible: false})}
                    onCancel={() => this.setState({editModalVisible: false})}
                    todoId={todo.id}
                    defaultDate={todo.notificationDate || new Date().getTime()}
                    defaultTime={todo.notificationTime || new Date().getTime()}
                    defaultDescription={todo.description}
                />
            </div>
        );
    }
}
