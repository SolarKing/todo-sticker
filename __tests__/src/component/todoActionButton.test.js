// @flow
import React from 'react';
import { shallow, type ShallowWrapper } from 'enzyme';
import TodoActionButton from '../../../src/component/todoActionButton';
import { Dropdown, Menu } from 'antd';
import { ipcRenderer } from 'electron';

jest.mock('electron', () => {
    let ipcRenderer = {
        send: jest.fn()
    };
    return {
        ipcRenderer
    };
});

describe('TodoActionButton', () => {
    let wrapper: ShallowWrapper;

    beforeAll(() => {
        wrapper= shallow(
            <TodoActionButton
                todo={{
                    id: '1',
                    dueDatetime: 10,
                    description: 'Hello!',
                    complete: false,
                    createTime: 0,
                }}
                enableEdit={true}
                enableComplete={true}
            />
        );
    });

    it('should render correct with complete and edit features are enabled', () => {
        wrapper.setProps({
            enableEdit: true,
            enableComplete: true,
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should render correct with complete feature is enabled', () => {
        wrapper.setProps({
            enableEdit: false,
            enableComplete: true,
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should render correct with edit feature is enabled', () => {
        wrapper.setProps({
            enableEdit: true,
            enableComplete: false,
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should render correct with complete and edit features are disabled', () => {
        wrapper.setProps({
            enableEdit: false,
            enableComplete: false,
        });
        expect(wrapper).toMatchSnapshot();
    });

    it('should execute ipc sender method', () => {
        let menu = shallow(wrapper.find(Dropdown).prop('overlay'));
        menu.find(Menu.Item).at(0).find('a').at(0).simulate('click');
        expect(ipcRenderer.send.mock.calls.length).toEqual(1);
        expect(ipcRenderer.send.mock.calls[0][0]).toEqual('addNote');
        expect(ipcRenderer.send.mock.calls[0][1]).toEqual({
            id: '1',
            noteDescription: 'Hello!',
        });
    });

});