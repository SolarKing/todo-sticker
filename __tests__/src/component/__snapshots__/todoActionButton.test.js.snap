// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoActionButton should render correct with complete and edit features are disabled 1`] = `
<div>
  <Dropdown
    mouseEnterDelay={0.15}
    mouseLeaveDelay={0.1}
    overlay={
      <Menu
        className=""
        prefixCls="ant-menu"
        theme="light"
      >
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Add Note
          </a>
        </MenuItem>
        <MenuItem>
          <Connect(DeleteTodoButton)
            todoId="1"
          />
        </MenuItem>
      </Menu>
    }
    placement="bottomLeft"
    prefixCls="ant-dropdown"
    trigger={
      Array [
        "click",
      ]
    }
  >
    <Button
      ghost={false}
      loading={false}
      prefixCls="ant-btn"
    >
      <Icon
        type="ellipsis"
      />
    </Button>
  </Dropdown>
  <Connect(EditModal)
    defaultDescription="Hello!"
    defaultDueDatetime={10}
    onCancel={[Function]}
    onSave={[Function]}
    title="Edit Todo"
    todoId="1"
    visible={false}
  />
</div>
`;

exports[`TodoActionButton should render correct with complete and edit features are enabled 1`] = `
<div>
  <Dropdown
    mouseEnterDelay={0.15}
    mouseLeaveDelay={0.1}
    overlay={
      <Menu
        className=""
        prefixCls="ant-menu"
        theme="light"
      >
        <MenuItem>
          <Connect(CompleteTodoButton)
            todoId="1"
          />
        </MenuItem>
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Edit
          </a>
        </MenuItem>
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Add Note
          </a>
        </MenuItem>
        <MenuItem>
          <Connect(DeleteTodoButton)
            todoId="1"
          />
        </MenuItem>
      </Menu>
    }
    placement="bottomLeft"
    prefixCls="ant-dropdown"
    trigger={
      Array [
        "click",
      ]
    }
  >
    <Button
      ghost={false}
      loading={false}
      prefixCls="ant-btn"
    >
      <Icon
        type="ellipsis"
      />
    </Button>
  </Dropdown>
  <Connect(EditModal)
    defaultDescription="Hello!"
    defaultDueDatetime={10}
    onCancel={[Function]}
    onSave={[Function]}
    title="Edit Todo"
    todoId="1"
    visible={false}
  />
</div>
`;

exports[`TodoActionButton should render correct with complete feature is enabled 1`] = `
<div>
  <Dropdown
    mouseEnterDelay={0.15}
    mouseLeaveDelay={0.1}
    overlay={
      <Menu
        className=""
        prefixCls="ant-menu"
        theme="light"
      >
        <MenuItem>
          <Connect(CompleteTodoButton)
            todoId="1"
          />
        </MenuItem>
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Add Note
          </a>
        </MenuItem>
        <MenuItem>
          <Connect(DeleteTodoButton)
            todoId="1"
          />
        </MenuItem>
      </Menu>
    }
    placement="bottomLeft"
    prefixCls="ant-dropdown"
    trigger={
      Array [
        "click",
      ]
    }
  >
    <Button
      ghost={false}
      loading={false}
      prefixCls="ant-btn"
    >
      <Icon
        type="ellipsis"
      />
    </Button>
  </Dropdown>
  <Connect(EditModal)
    defaultDescription="Hello!"
    defaultDueDatetime={10}
    onCancel={[Function]}
    onSave={[Function]}
    title="Edit Todo"
    todoId="1"
    visible={false}
  />
</div>
`;

exports[`TodoActionButton should render correct with edit feature is enabled 1`] = `
<div>
  <Dropdown
    mouseEnterDelay={0.15}
    mouseLeaveDelay={0.1}
    overlay={
      <Menu
        className=""
        prefixCls="ant-menu"
        theme="light"
      >
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Edit
          </a>
        </MenuItem>
        <MenuItem>
          <a
            onClick={[Function]}
          >
            Add Note
          </a>
        </MenuItem>
        <MenuItem>
          <Connect(DeleteTodoButton)
            todoId="1"
          />
        </MenuItem>
      </Menu>
    }
    placement="bottomLeft"
    prefixCls="ant-dropdown"
    trigger={
      Array [
        "click",
      ]
    }
  >
    <Button
      ghost={false}
      loading={false}
      prefixCls="ant-btn"
    >
      <Icon
        type="ellipsis"
      />
    </Button>
  </Dropdown>
  <Connect(EditModal)
    defaultDescription="Hello!"
    defaultDueDatetime={10}
    onCancel={[Function]}
    onSave={[Function]}
    title="Edit Todo"
    todoId="1"
    visible={false}
  />
</div>
`;
